### PR TITLE
Camera system (issue 44)

### PR DESCRIPTION
**Pull request by Davidek523**
# Pull Request Template

# Code Pull Request Template

## Summary

This PR  adds camera system and modifies player's movement  This PR adds a functional camera system to the game. It means that the screen will follow the player everywhere they move. I achieved this by adding a new class in sprites.py named AllSprites. It's a sprite group class that takes all the sprites in the sprite group and implements the camera logic on them. 

For this I added the Player class to the sprite group and changed the all_sprite variable in the gameloop. It uses now the AllSprites class instead of making the sprite group directly in gameloop. 

I also added a new movement method in the Player class and commented the old one. When debuggin the camera I experienced the original movement method slow and quite static with a couple of bugs so I used a movement that uses WSAD keys as input. This way the movement was quick and smooth in my experience. 

## Related Issues
 

This PR  Closes Issue #44  

---

## ✅ Checklist:

  - [x] 🛠️ I have tested my code, and it works as expected (e.g., game runs without errors).
  - [x] 📝 I have made sure that the code follows the project's formatting and style guidelines.

Optional:
  - [x] 💬 Code is commented (if applicable)

---

## 🏷️ Labels:

  - `type: feature` 
  - `area: gameplay`
  - `area: assets`
  - `game: mechanics`

### CODE REVIEW

**Comment by ultimateownsz** [comment]
Hey @Davidek523 you need to use the correct pr template, because I'm not sure what this change does! You have to make another PR with the right PR template chosen!

**Comment by ultimateownsz** [comment]
I reopened your PR, you have to copy and paste the PR template you need to make your changes more clear! :)

File: src/sprites.py [hunk 1]
```diff
@@ -1,101 +1,158 @@
...
-            else:
-                # go left
-                self.player_preview_rect.x = self.rect.x - TILE_SIZE
-        # move the gost on the y axis
+    #     self.rect: pygame.Rect = self.image.get_rect()
+    #     # keep track of the transparent preview of the next move
+    #     self.player_preview_rect = self.rect.copy()
+    #     # this is used to only move once when the mouse is pressed
+
+    '''DEBUGGING MOVEMENT METHOD'''
```
Comment (ultimateownsz) [inline src/sprites.py:57]: I think for debugging purposes, this WASD movement implementation is fine and functional, especially for testing basic player movement logic. However, I’d like to highlight a few concerns:

1. **Mouse interaction in the game loop:**
- Mouse interaction is a key feature of the game, as it drives multiple actions (movement, interacting with objects)
- By removing mouse-based interactions and replacing them with WASD movement, we lose a significant part of the game's intended design.

2. **Ghost preview visualization:**
- The shadow opacity method (from @PurpleProg implementation) effectively acts as a **preview of where the player can go**. This visualization is crucial for providing feedback to the player before committing to a move.
- WASD movement does not provide any visualization, which makes it harder for players to understand their options or plan movements, especially if we do add a turn base system into the player movement.
- The ghost system is not just a visual feature but an integral part of the gameplay, aligning with strategies like games as Civilization or Unciv.
- Example of how it could look like, our game is grid based so I could not find the best representation: 
unciv player movement


3. **Representation of the gameplay:**
- While WASD movement may simplify debugging, it diverges significantly from the game's core mechanics as describes in the GGD.
- A ghost preview system reinforces the idea of a deliberate, turn-based approach, providing clarity and preventing accidental moves.

4. **Suggestion**:
- If the WASD movement is intended purely for debugging, consider isolating it behind a debug flag or toggle, making sure that the primary interaction system (mouse-based with ghost preview) remains intact and active during gameplay. 
- Maintaining the visualization aspect alongside debugging tools would better represent the intended gameplay experience.

In summary, while the WASD movement approach serves a purpose for debugging, it doesn’t align with the visualization-first approach of the game. Retaining the mouse interaction and ghost preview in the main loop is essential for representing the core mechanics and player experience.

**Review by ultimateownsz — CHANGES_REQUESTED** [review]
I left a comment about what could be altered in the `src/GUI/gameloop.py` file

**Review by ultimateownsz — APPROVED** [review]
I tested it, it works great!
